### ISSUE TITLE
Add recursive option

### ISSUE THREAD

**Issue by styfle** [lycheeverse/lychee#78]
It would be nice to pass a URL and have it crawl the entire website recursively looking for dead links.

In order to avoid crawling the entire internet, it should stop recursing once a request no longer matches the original domain.

**Comment by mre** [comment]
Yeah it has been discussed a few times already and it's high on the todo-list. 
Will require quite some restructuring I guess.
Right now the flow is

```
main -> extractor -> channel -> client (link checker) -> main
```

but there is no connection back to the extractor.

**Comment by frederickjh** [comment]
The lack of recursive spidering make this project unusable for my purpose of checking all links internal and external on a website. I am trying to find a replacement for  [michaeltelford/broken_link_finder ](https://github.com/michaeltelford/broken_link_finder). It is written in Ruby and with out super user access a the new place where this will run it is impossible to install.  I am looking for a "portable" replacement. I worked with michaeltelford to get his project into a much more usable state. Check out that projects issue queue for some of the reasoning that went into the development.

In any case regarding this issue. The spidering should stop with links found on the current domain, but the links found to external sources should still be checked.

**Comment by mre** [comment]
#165 is getting very close to completion. It implements the functionality described. If you like to support, please build the version from that branch and test it. Feedback on the pull request is appreciated.

**Comment by frederickjh** [comment]
@mre I am new to rust but it seemed pretty straight forward as to how to build from [Working on an Existing Cargo Package](https://doc.rust-lang.org/cargo/guide/working-on-an-existing-project.html). However I have run into an issue. At first I thought it was a credential issue but it looks like a 404 issue.

![image](https://user-images.githubusercontent.com/864159/112447909-2145f680-8d52-11eb-81eb-cac2454cca6c.png)

> Caused by:
>  Unable to update https://github.com/amaurym/async-smtp?branch=am-fast-socks#eac57391

That GitHub URL returns a 404 so I am not sure how to proceed to build.

Rust information:
```shell
$ rustc --version
rustc 1.50.0 (cb75ad5db 2021-02-10)
$ rustup --version
rustup 1.23.1 (3df2264a9 2020-11-30)
info: This is the version for the rustup toolchain manager, not the rustc compiler.
info: The currently active `rustc` version is `rustc 1.50.0 (cb75ad5db 2021-02-10)`
$ cargo --version
cargo 1.50.0 (f04e7fab7 2021-02-04)
```

The GitHub repository **amaurym/async-smtp** does not seem to exist anymore.
Not sure how to proceed. Please advise. Thanks!

**Comment by frederickjh** [comment]
I dug through the `Cargo.lock` file for  [reacherhq/check-if-email-exists](https://github.com/reacherhq/check-if-email-exists) and found [this line](https://github.com/reacherhq/check-if-email-exists/blob/bbf218ab5e8f379312d922763ce4c87dd66f4a69/Cargo.lock#L140) with the source for `async-smtp`

```rust
source = "git+https://github.com/async-email/async-smtp?branch=master#0f1c4c6a565833f8c7fc314de84c4cbbc8da2b4a"
```
So it looks like the source for async-smtp has moved to [async-email/async-smtp](https://github.com/async-email/async-smtp).

**Comment by frederickjh** [comment]
Just found #189 which is the same issue I reported here as to why the build fails.

**Comment by mre** [comment]
Yeah. Related:
* https://github.com/lycheeverse/lychee/issues/184
* https://github.com/reacherhq/check-if-email-exists/issues/888

We are blocked by upstream at the moment. 😕

**Comment by frederickjh** [comment]
Ah, upstream [reacherhq/check-if-email-exists](https://github.com/reacherhq/check-if-email-exists) updated to use upstream instead of his fork three days ago. I am guessing he also deleted the fork then, but this project is still using it.

See here: [chore: Update wording around licenses #892](https://github.com/reacherhq/check-if-email-exists/pull/892/commits/47c1dd4e9e9eae7d12c635f6871049643492da4d#)

This repository still has references to his fork that no longer exists on both the `master` and `simple-recursion` branches in the `Cargo.lock` files. There is a comment in the `Cargo.toml` file that says:

    # Switch back to version on crates.io after
    # https://github.com/async-email/async-smtp/pull/36
    # is merged and a new version of check-if-email-exists is released

So it looks like pull 36 in the upstream is closed but the new crate has not been published as the newest one is dated January 10.

@mre let me know if there is any movement on this and I will then try to build from the `simple-recursion` branch and test.

**Comment by frederickjh** [comment]
[Upstream is reporting](https://github.com/reacherhq/check-if-email-exists/issues/888#issuecomment-811371149):
> This is fixed in 0.8.21

but I still cannot bulid from the `simple-recursion` branch, so I think something needs work there too be for this will build.

**Comment by frederickjh** [comment]
So, I think that the version of `async-email/async-smtp` needs to be upgraded from 0.8.19 to **0.8.21** for the `simple-recursion` branch to build.

**Comment by mre** [comment]
Thanks for the info. I'll tackle that once #208 is merged. 😄

**Comment by frederickjh** [comment]
@mre I see that #208 got merged back in April. Let me know if you get this branch to the point where it will build and I can then test it.

**Comment by frederickjh** [comment]
@mre I am still willing to test this but I will be finished with my current job in second week of June and may not have a need for it after that for a while. I would like to get this setup to replace the current program that we are using to check for broken links that I cannot easily move to a shared server because it is ruby. Let me know if you get the version of  `simple-recursion` changed so that this branch builds and I will test it.

**Comment by mre** [comment]
Thanks for your patience. Want to work on this as soon as I find the time. No guarantees this will be soon, though. 😅

**Comment by frederickjh** [comment]
@mre Patience I have, but time is running out. I finish at my current work place on June 9. I had hope to use this to replace a Ruby broken link checker that is running on an in-house server I need to decommission. I need something I can run on shared hosting with out needing to install a bunch of dependencies I don't have permission to do so.
So, if you could find a little time this week to get the branch in a shape that it will build, I could build and test it .

**Comment by untitaker** [comment]
I found that [muffet](https://github.com/raviqqe/muffet) and [linkcheck](https://github.com/filiph/linkcheck) serves the recursive usecase the best. What neither does is to opportunistically check /sitemap.xml to traverse the site faster/get to efficient parallelization faster. Lychee could one-up them on performance if that is done by default.

**Comment by mre** [comment]
New PR which tackles this: #465 
Will probably go through another round of refactoring before it's ready, but I'm on it.

**Comment by cipriancraciun** [comment]
I'm unsure how this is actually implemented, perhaps what I am about to say is already covered, so sorry for the duplication.

Recursion is also very important to me, but I would like to allow the user to specify a list of origins (scheme+host+port) to allow recursion for, or a list of regular expressions.

Say for example one has both a `www` and a `blog`, site but also `docs` site, one would like to primarily check the `www` and `blog` (thus specifying them as arguments), but also to recursively check everything that links towards the `docs` site and other pages from that on any of the three.

**Comment by mre** [comment]
Good point. It's not implemented and wasn't mentioned before.

The way I envisioned it was that all links which belong to the same input URI would be followed recursively, while the rest would not. So you could do `lychee --recursive www blog docs`, but it sounds like you only want to check that links pointing to `docs`, but not all of `docs`. I wonder what's the issue to check all links in `docs`, though. Is the site too big?
If you want to exclude some URI patterns for `docs`, you could do `lychee --exclude docs/foo --recursive www blog docs`.

**Comment by cipriancraciun** [comment]
> So you could do `lychee --recursive www blog docs`, but it sounds like you only want to check that links pointing to `docs`, but not all of `docs`. I wonder what's the issue to check all links in `docs`, though. Is the site too big?

Imagine that instead of `docs` there is actually an assets domain, that might not even have an index to start from;  however this assets domain could contain some HTML files that are perhaps included in `<iframe>` inside `www` or `blog`, and some of these HTML files are somewhat self-contained, thus starting from them one wouldn't reach the entire assets collection.  Now if one of these HTML files actually contains broken links, that could affect the initial `www` and `blog` sites.

Or another example, that extra domain could be something hosting examples HTMLs that might be linked from the main site, and one would like to make sure that every example works as expected.

----

Or if the above reasons don't seem convincing enough (granted they are quite extreme), I assume that inside the code there already exists a set of "allowed" domains or origins for recursion, that are filled in at startup based on the starting links;  thus allowing the user to manipulate that wouldn't be much of a burden, but will also increase flexibility.

**Comment by mre** [comment]
Hm... the main question is always how to wire that up in the CLI without provoking additional mental overhead.
Following our `--exclude`/`--include` patterns, we could add an `--include-recursive` parameter:

```sh
lychee --recursive --include-recursive docs -- www blog docs
```

**Comment by frittentheke** [comment]
@mre could you maybe give an update on support for traversing / recursive link checking to tackle a whole website?

**Comment by mre** [comment]
Yes, sure.

There were a few attempts, but there were always issues with the design. It's a feature which touches on almost all parts of the code and we have to get this right.

I'd love to dedicate more time to it,  but it's hard to add that feature next to other responsibilities. Currently looking into companies who might be willing to sponsor that feature as I guess it will be quite some work, but it would have a very positive impact to the usefulness for all users. I know that there are companies out there which would really like to have that, but so far there hasn't been a lot of traction with regards to sponsoring.
My hope is to still get the free time to work on it at some point, but I wouldn't hold my breath for it right now unless there's a way to fund this. In the meantime, I encourage others to take a stab at it as well.

**Comment by styfle** [comment]
I'll close this since I already built a solution.

https://github.com/styfle/links-awakening

https://www.npmjs.com/package/links-awakening

**Comment by mre** [comment]
Nice package.
I would still like to keep the issue open, as I'd like to add recursion support to lychee at some point as well.

**Comment by Alseenrodelap** [comment]
Still no recursive option in the main branch since 2020? I'm trying to run this great program via Docker but really miss the recursive option...

**Comment by lfrancke** [comment]
I'm happy to offer a bounty of sorts of 100€ (payable via PayPal or SEPA) payment for whoever implements this, if multiple people work on it I'm happy to split the money.

I know this won't cover the whole development of this feature.

**Comment by ewen-lbh** [comment]
@lfrancke hiya, does your offer still stand?

**Comment by lfrancke** [comment]
Thanks for checking. Yes, it does.

**Comment by lukehsiao** [comment]
Somewhat tangential: I'm a big fan of `lychee`, and my workaround is to just link check all the links in a sitemap.

I have a tool for this: https://github.com/lukehsiao/sitemap2urllist

```
sitemap2urllist https://www.numbersstation.ai/sitemap.xml --cache | xargs lychee --cache
```

which seems to serve pretty well.

Likely will be obsolete if/once lychee adds recursive support, but useful in the meantime.

**Comment by ewen-lbh** [comment]
> Thanks for checking. Yes, it does.

I saw that even people familiar with the codebase failed after 3 attempts because of design issues so i must admit i'm a bit intimidated haha '^^

I guess a massively parallel recursive walk is the goal? @mre , was the v3 attempt closed because of a design issue, or just because it had bugs still and had drifted too much from main?

**Comment by mre** [comment]
Yeah, the third attempt changed too many things at the same time; I ran into issues and let the branch diverge too much, which made progress harder.

In my opinion, the first attempt, while simplistic, had the best chance to get merged. With that first attempt, I ran into a weird edge case where the pipeline would not terminate. It simply got stuck somewhere. Probably because my count of outstanding requests was off.

If you are willing to give it a try, feel free to take a look at the different versions to see which one you like best and pick up the work from there. Of course, a "clean room" implementation might also work. In fact, too much familiarity with the codebase might be a hindrance for finding a good solution. 

In any case, thanks for looking into that! Good luck and if you have any questions, feel free to reach out here or send me an email so that we can chat about ideas or brainstorm a bit.

**Comment by ewen-lbh** [comment]
Ok so I implemented something, it's more of a PoC for now but it works

I just uuuuuh left implementation of a recursion depth limit for later and I, for funsies, tried on en.wikipedia.org and then std.rust-lang.org, with `--max-concurrency 4096` ... currently on mobile internet data sharing with my wifi box rebooting because i literally kicked myself off the internet (literally, spotify did not load anymore, couldn't ssh, couldn't push to github etc)

the implementation might be a little too "efficient" lmao

**Comment by ewen-lbh** [comment]
theres also an issue with tokio never unlocking sends on the responses channel once they get filled up, a band aid is to bump `--max-concurrency` but it's not a real solution

**Comment by mre** [comment]
Great!

> the implementation might be a little too "efficient" lmao

haha. 😆 

> a band aid is to bump --max-concurrency but it's not a real solution

You're right. Simply increasing max-concurrency is indeed just masking the underlying issue by giving more room before it manifests, rather than addressing why the channel isn't properly draining in the first place.

I wonder if it's just backpressure.

**Comment by almereyda** [comment]
With [Lychee being awarded an nlnet grant](https://nlnet.nl/news/2025/20250321-call-announcement-core.html), may we expect this feature to be in scope for this next iteration?

**Comment by mre** [comment]
Yes. 😊 We will still do an official announcement, but recursion support is one objective. Right now we're in the planning phase for the tasks.

**Comment by 6543** [comment]
one implementation detail should be to limit the recursion to a domain/base-url list to e.g. only check mysite.com/test* and an option to allow to check one pepth more outside of the basie-url ...

e.g. if mysite.com/test/some.html links to external sites to check it but dont add it to the recursion to get more sites to check ...

hope that makes sense and is not already mentioned somewhere :)

my usecase would be to itterate over a full website, that one do own/maintain to find dead links in there *somewhere* ...

**Comment by mre** [comment]
Yes, exactly, that is the plan. We limit recursion to input domains by default.

```
lychee https://endler.dev
```

Would do the "right thing" and only handle links on my blog.

**Comment by ricardbejarano** [comment]
@mre when can we expect this to be released?

**Comment by mre** [comment]
@ricardbejarano, the best way forward is to review https://github.com/lycheeverse/lychee/pull/1603 and perhaps test it to see if it works for real-world scenarios. That would really help because at the moment we are a bit resource-constrained, but we would love to push that over the finishing line.

Don't worry too much about the file conflicts; you should be able to test it with

```sh
cargo install lychee --git http://github.com/gwennlbh/lychee --branch recursion
lychee -R https://example.org --recursed-domains example.org 
```

**Comment by lkwg82** [comment]
any progress on this issue?

**Comment by mre** [comment]
Work has been happening on https://github.com/lycheeverse/lychee/issues/1898, which discusses architectural changes that will make recursion support a by-product of that refactor. We'd appreciate any input there, although the discussion is about intricate details of lychee's design, which could be a bit harder to grasp without additional knowledge.

We support sitemaps now as part of #2062, which is a useful workaround for many websites.

We unified request handling in https://github.com/lycheeverse/lychee/issues/1886. This way, CLI inputs and links discovered at runtime are treated the same. That unification helps with adding recursion support in the future.

There are a few more bits and pieces we merged that get us closer to recursion support, but it's still a long way to go.

**Comment by mre** [comment]
Actually, I totally forgot to mention the most important recent development, which is the `WaitGroup` primitive by @katrinafyi, merged in PR #2046. This directly addresses the termination problem that defeated every prior attempt.

`WaitGroup` is an RAII-based mechanism for waiting on a dynamic set of tasks that can spawn other tasks. It consists of:

- `WaitGroup`: A single waiter that detects when all work is done, and
- `WaitGuard`: A clonable guard held by each task. When all guards are dropped, the waiter completes.

The key insight is that `WaitGuard` can be cloned, allowing a task to spawn sub-tasks (recursion!) while maintaining the invariant that the `WaitGroup` only completes when *all* guards (including those held by recursive sub-tasks) are dropped.

This solves our termination problem:

```rust
let (waiter, guard) = WaitGroup::new();

// Each request carries a guard clone
send_req.send((guard.clone(), request)).await;

// In the response handler, if recursing:
// The guard is cloned for each new request
for new_request in discovered_links {
    send_req.send((guard.clone(), new_request)).await;
}

// The original guard is dropped when the response is fully processed.
// When ALL guards are dropped (no more work), waiter.wait() returns.
```

The `WaitGroup` is already integrated into lychee's main check loop. The `collect_responses` function uses `take_until(waiter.wait())` to stop receiving when all work is done. The current code even has a comment anticipating recursion:

https://github.com/lycheeverse/lychee/blob/8bbeeb57f3055a97b27495610919ff1169b963f9/lychee-bin/src/commands/check.rs#L203-L207

This is the missing piece that all four previous attempts (#165, #465, #489, #1603) lacked.

Another important one is  per-host rate limiting (#1929), which ensures that recursive crawling respects server limits. This removes a major blocker that was previously deemed "out of scope" but was in practice essential for real-world recursion.

**Comment by mre** [comment]
I think the biggest meta-lesson for me is that recursion cannot be landed as a single massive PR. The successful path is to land prerequisites incrementally. We are already on a good trajectory by landing functionality like the following:

- WaitGroup for dynamic task termination
- Unified request handling through HostPool
- Per-host rate limiting
- Sitemap support as a discovery mechanism

Each of these were reviewed, tested, and merged independently, which helped us keep up the momentum.
The project is now in the best position it's ever been. The `WaitGroup` solves termination. Unified host pools solve configuration consistency. Per-host rate limiting prevents accidental DDoS. We should continue down that path as it has proven to be the most promising so far.

**Comment by 6543** [comment]
In this regard, what would be the rough roadmap till recursion?
so to let people know what to focus on if interested in this issue ;)

**Comment by almereyda** [comment]
To me your question is answered in the post before yours.

**Comment by mre** [comment]
> what would be the rough roadmap till recursion?

I cannot give you a roadmap for recursion at the moment, but there's ongoing work here that we'd like to get feedback on:

* #2097 
* #1898 
* #2129

**Comment by tomashofmanng** [comment]
@lfrancke, does your 100 EUR PayPal/SEPA offer still stand in 2026? Since the maintainers prefer incremental prerequisites rather than another large recursion PR, would you also be willing to split the bounty for a focused contribution that is explicitly accepted as advancing recursion support?

@mre, if so, could you point me to the smallest currently unclaimed prerequisite or acceptance gap worth tackling? I noticed PR #2229 overlaps the channel-lifetime design, so I want to avoid duplicating active work.

**Comment by lrt950** [comment]
I implemented the recursive website checker in #2283. It uses the existing `WaitGroup`/shared `HostPool` groundwork, adds domain and depth controls, and includes regression coverage for the bounded-channel deadlock, external domains, and redirects.

@lfrancke, this is intended to cover the full feature from your 100 EUR bounty offer. If the maintainers accept and merge it, I can coordinate PayPal payout details privately.

**Comment by lfrancke** [comment]
Hi, sorry for the late response: Yes, the offer still stands. I'm happy to split it up as well, yes.

**Comment by FloatingPegasus** [comment]
I reproduced a separate depth-limit correctness issue in #2283: if a longer route reaches a shared page before a delayed shorter route, the crawler can exit successfully while missing a broken descendant inside `--max-depth`.

The focused fix and local regression are submitted to the feature branch at https://github.com/lrt950/lychee/pull/2. The regression fails on the current feature branch and on the companion deduplication implementation; with this patch all seven local recursion CLI tests pass, together with Clippy and formatting checks. The PR documents the integration needed if the deduplication patch lands first.

@lfrancke, thanks for reconfirming the shared reward. If this contribution is accepted into the completed feature, would a €20 share be available for this depth-correctness work? I understand the €100 is the total pool shared with the feature author and other contributors, and that no share is agreed yet.

**Comment by sen-ye** [comment]
@lfrancke, thank you for reconfirming the shared €100 bounty. I submitted a distinct companion fix at https://github.com/lrt950/lychee/pull/3: recursive content-fetch failures now fail the CLI, and malformed links remain visible in the error summary while valid links are still checked.

The PR contains two local regressions, eight passing recursive tests, and passing all-target/all-feature Clippy. It also records the external-network failures in the broader local and public CI suites.

If this patch is accepted into #2283 and the completed recursive feature lands upstream, would a €20 share of the existing pool be available through PayPal? I understand that the allocation is still your decision. Payment details can be exchanged privately after confirmation.

OpenAI Codex authored and tested this contribution on behalf of @sen-ye.

**Comment by lfrancke** [comment]
I have no idea where all this activity is coming from all of a sudden. I'm sorry @mre, I didn't mean to attract unwanted PRs/work.
I was also contacted via mail. Please stop that. I am happy to honor my bounty for whenever this is implemented at my very own discretion. I don't know anything about the details that need to get done to implement this and I don't want to deal with the details either. This was meant as a well intentioned incentive years ago. I don't know how many people there are behind these accounts and I don't know wheret his sudden interest comes from.

**Comment by sen-ye** [comment]
Sorry for adding to the noise and creating extra coordination work. I have withdrawn my bounty-share request and closed my companion PR at https://github.com/lrt950/lychee/pull/3. I will not contact you further about payment. Thank you for clarifying.
